core/src/banking_simulation.rs

```diff
@@ -0,0 +1,616 @@
+// block's hashes would definitely differ than the recorded ones as slight block composition
+// difference is inevitable.
+//
+// For poh time, we just use PohRecorder as same as the real environment, which is just 400ms
+// timer, external to banking stage and thus mostly irrelevant to banking stage performance. For
+// wall time, we use the first BankStatus::BlockAndBankHash and `SystemTime::now()` to define T=0
+// for simulation. Then, simulation progress is timed accordingly. As a context, this syncing is
+// needed because all trace events are recorded in UTC, not relative to poh nor to leader schedule
+// for simplicity at recording.
+//
+// Lastly, here's the last and most complicated input to simulate: transactions.
+//
+// A bit closer look of transaction load profile is like below, regardless of internal banking
+// implementation and simulation:
+//
+// There's ever `BufferedPacketsDecision::Hold`-ed transactions to be processed as the first leader
+// slot nears. This is due to solana's general tx broadcast strategy of node's forwarding and
+// client's submission, which are unlikely to chabge soon. So, we take this as granted. Then, any
+// initial leader block creation starts with rather large number of schedule-able transactions.
+// Also, note that additional transactions arrive for the 4 leader slot window (roughly ~1.6
+// seconds).
+//
+// Simulation have to mimic this load pattern while being agnostic to internal bnaking impl as much
+// as possible. For that agnostic objective, `TracedSender`s are sneaked into the SigVerify stage
+// and gossip subsystem by `BankingTracer` to trace **all** of `BankingPacketBatch`s' exact payload
+// and _sender_'s timing with `SystemTime::now()` for all `ChannelLabel`s. This deliberate tracing
+// placement is not to be affected by any banking-tage's capping (if any) and its channel
+// consumption pattern.
+//
+// BankingSimulator consists of 2 phases chronologically: warm-up and on-the-fly. The 2 phases are
+// segregated by the aforementioned T=0.
+//
+// Both phases just sends BankingPacketBatch in the same fashion, pretending to be sigveirfy
+// stage/gossip while burning 1 thread to busy loop for precise T=N at ~1us granularity.
+//
+// Warm up is defined as T=-N secs using slot distance between immediate ancestor of first
+// simulated block and root block. As soon as warm up is initiated, we invoke
+// `BankingStage::new_num_threads()` as well to simulate the pre-leader slot's tx-buffering time.
+pub struct BankingSimulator {
+    banking_trace_events: BankingTraceEvents,
+    first_simulated_slot: Slot,
+}
+
+#[derive(Error, Debug)]
+pub enum SimulateError {
+    #[error("IO Error: {0}")]
+    IoError(#[from] io::Error),
+
+    #[error("Deserialization Error: {0}")]
+    SerializeError(#[from] bincode::Error),
+}
+
+pub struct BankingTraceEvents {
+    packet_batches_by_time: BTreeMap<SystemTime, (ChannelLabel, BankingPacketBatch)>,
```

Comment:
Why not use a `VecDeque` here? These are, afaik, always in order from the files (assuming we read the files in the correct order, which is easy enough to do).

Reply:
added comments: d3bf0d9c60c85acfc0df38d4bda52b0feb98bbc1